### PR TITLE
Authorization Improvements

### PR DESCRIPTION
**Pull request by tommoor**
Moving to a more consistent authorization methodology, fixes some minor holes in the process and add regression specs.

### CODE REVIEW

File: server/api/collections.js [hunk 1]
```diff
@@ -32,13 +34,27 @@ router.post('collections.create', auth(), async ctx => {
...
+});
+
 router.post('collections.update', auth(), async ctx => {
   const { id, name, color } = ctx.body;
   ctx.assertPresent(name, 'name is required');
   if (color)
     ctx.assertHexColor(color, 'Invalid hex value (please use format #FFFFFF)');
 
   const collection = await Collection.findById(id);
+  authorize(ctx.state.user, 'update', collection);
```
Comment (tommoor) [inline server/api/collections.js]: 🍖 Ensure that the user is authorized to update the collection

File: server/api/index.js [hunk 2]
```diff
@@ -38,6 +38,16 @@ api.use(async (ctx, next) => {
       }
     }
 
+    if (message.match('Authorization error')) {
```
Comment (tommoor) [inline server/api/index.js]: This catches the error that's thrown by `authorize()` if the user does not have correct permissions.

I'm trying to match the pattern we had before here, which is return a `404` if attempting to read a document you don't have access to and a `401` if any other action.

File: server/api/middlewares/authentication.js [hunk 3]
```diff
@@ -96,10 +95,6 @@ export default function auth(options: AuthOptions = {}) {
         }
       }
 
-      if (options.adminOnly && !user.isAdmin) {
```
Comment (tommoor) [inline server/api/middlewares/authentication.js]: Moved this logic to `policies/`

File: server/api/team.js [hunk 4]
```diff
@@ -31,41 +32,41 @@ router.post('team.users', auth(), pagination(), async ctx => {
...
-  if (!promotedUser) throw httpErrors.NotFound();
+  const user = await User.findById(userId);
+  authorize(ctx.state.user, 'promote', user);
 
-  await team.addAdmin(promotedUser);
+  const team = await Team.findById(teamId);
+  await team.addAdmin(user);
 
-  ctx.body = presentUser(ctx, promotedUser, { includeDetails: true });
+  ctx.body = {
```
Comment (tommoor) [inline server/api/team.js]: Updated the response here, this was missing in the user data with other parts of the response. These endpoints seem to be the only place where a `data` property wasn't used.

File: server/api/views.js [hunk 5]
```diff
@@ -1,24 +1,26 @@
...
+const { authorize } = policy;
 const router = new Router();
 
 router.post('views.list', auth(), async ctx => {
   const { id } = ctx.body;
   ctx.assertPresent(id, 'id is required');
 
+  const user = ctx.state.user;
+  const document = await Document.findById(id);
+  authorize(user, 'read', document);
```
Comment (tommoor) [inline server/api/views.js]: Previously there was no check here - you could record a "view" against any document in the db if you know the `id`. `views.list` would then happily return that document.

File: server/api/views.js [hunk 6]
```diff
@@ -42,11 +44,13 @@ router.post('views.create', auth(), async ctx => {
...
   const user = ctx.state.user;
   const document = await Document.findById(id);
-
-  if (!document || document.teamId !== user.teamId)
-    throw httpErrors.BadRequest();
+  authorize(user, 'read', document);
 
   await View.increment({ documentId: document.id, userId: user.id });
+
+  ctx.body = {
```
Comment (tommoor) [inline server/api/views.js]: Just trying to be more consistent with our API responses.

File: server/models/Team.js [hunk 7]
```diff
@@ -42,13 +42,13 @@ Team.prototype.createFirstCollection = async function(userId) {
...
 Team.prototype.addAdmin = async function(user: User) {
-  return await user.update({ isAdmin: true });
+  return user.update({ isAdmin: true });
 };
 
 Team.prototype.removeAdmin = async function(user: User) {
   const res = await User.findAndCountAll({
     where: {
-      teamId: user.teamId,
+      teamId: this.id,
```
Comment (tommoor) [inline server/models/Team.js]: The controller was checking the authenticated users team, but then the model was using the promoted users team – these are the same right now, but makes sense to make these match.

File: package.json [hunk 8]
```diff
@@ -87,6 +87,7 @@
     "boundless-popover": "^1.0.4",
     "bugsnag": "^1.7.0",
     "bull": "^3.3.7",
+    "cancan": "3.1.0",
```
Comment (tommoor) [inline package.json]: I checked out the source code of this, it's nice and simple with a minimal API so we can move it in house without too many changes if it turns out that it's restrictive at all.

**Comment by jorilallo** [comment]
Comment on return codes:

- `401` when authentication token is invalid
- `404` when user tries to access invalid resource (or they are trying to access a resource they should not see)
- `403` when user tries to perform an action they aren't allowed to perform (e.g. non-admin tries to promote an admin)
- `400` invalid request, e.g. wrong params etc

**Comment by tommoor** [comment]
@jorilallo changes as discussed 🙏 . If you're down with this approach once this is merged there are a lot of places where we're calling `httpErrors.BadRequest` directly – we should strip all of those out in favor of specific error classes like `ValidationError` as each one is getting a unique error id based on the passed string.

File: server/api/__snapshots__/team.test.js.snap [hunk 9]
```diff
@@ -40,15 +41,16 @@ Object {
...
+  "message": "An admin role is required to access this resource",
   "ok": false,
+  "status": 403,
 }
 `;
 
 exports[`#team.removeAdmin shouldn't demote admins if only one available  1`] = `
 Object {
-  "error": "at_least_one_admin_is_required",
+  "error": "validation_error",
```
Comment (tommoor) [inline server/api/__snapshots__/team.test.js.snap]: This is an example if we replace `httpErrors.BadRequest` out in all the spots it's used - the error responses are more useable. I'll leave that for a separate PR though.

**Comment by jorilallo** [comment]
Looks great, thanks for this

**Pull request merged**